bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/ResultRowAdapter.java

```diff
@@ -0,0 +1,181 @@
+     * RowBuilder} has ordering as:
+     *
+     * <ul>
+     *   <li>family names clustered, but not sorted
+     *   <li>qualifiers in each family cluster is sorted lexicographically
+     *   <li>then descending by timestamp
+     * </ul>
+     *
+     * The end result will be that {@link RowCell} are ordered as:
+     *
+     * <ul>
+     *   <li>lexicographical by family
+     *   <li>then lexicographical by qualifier
+     *   <li>then descending by timestamp
+     * </ul>
+     *
+     * A flattened version of the {@link RowCell} map will be sorted correctly.
+     */
+    @Override
+    public void finishCell() {
+      if (!Objects.equals(this.family, this.previousFamily)) {
+        previousFamily = this.family;
+        currentFamilyCells = new ArrayList<>();
+        cells.put(this.family, this.currentFamilyCells);
+      }
+
+      RowCell rowCell =
+          new RowCell(
+              this.currentKey,
+              this.family.getBytes(),
+              ByteStringer.extract(this.qualifier),
+              TimestampConverter.bigtable2hbase(this.timestamp),
+              ByteStringer.extract(this.value),
+              this.labels);
+      this.currentFamilyCells.add(rowCell);
+      totalCellCount++;
+    }
+
+    /**
+     * This method flattens the {@code cells} which has a map of Lists keyed by family name. The
+     * {@code cells} treeMap is sorted lexicographically, and each List is sorted by qualifier in
+     * lexicographically ascending order, and timestamp in descending order.
+     *
+     * @return an object of HBase {@link Result}.
+     */
+    @Override
+    public Result finishRow() {
+      if (totalCellCount == 0) {
+        return Result.EMPTY_RESULT;
+      }
+
+      ImmutableList.Builder<Cell> combined = ImmutableList.builder();
+      for (List<RowCell> familyCellList : cells.values()) {
+
+        RowCell previous = null;
+        for (RowCell rowCell : familyCellList) {
+          if (previous == null || !rowCell.getLabels().isEmpty() || !keysMatch(rowCell, previous)) {
+            combined.add(rowCell);
+          }
+          previous = rowCell;
+        }
+      }
+
+      return Result.create(combined.build());
+    }
+
+    private boolean keysMatch(RowCell current, RowCell previous) {
+      return current.getTimestamp() == previous.getTimestamp()
+          && Arrays.equals(current.getQualifierArray(), previous.getQualifierArray());
+    }
+
+    @Override
+    public void reset() {
+      this.currentKey = null;
+      this.family = null;
+      this.qualifier = null;
+      this.labels = null;
+      this.timestamp = 0L;
+      this.value = null;
+      this.cells = new TreeMap<>();
+      this.currentFamilyCells = null;
+      this.previousFamily = null;
+      this.totalCellCount = 0;
+    }
+
+    @Override
+    public Result createScanMarkerRow(ByteString rowKey) {
+      return Result.EMPTY_RESULT;
```

Comment:
I'm a little concerned about re-using the EMPTY_RESULT marker here.

Can we create our own static instance of an empty Result and use referential equality to detect it as a marker?